### PR TITLE
Added ReadOnlySpan

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
This is mainly copy-paste and some small tweaks from the existing r/w implementation.

@krwq

### CODE REVIEW

**Comment by KrzysztofCwalina** [comment]
@dotnet-bot test this please

**Comment by KrzysztofCwalina** [comment]
@dotnet-bot test this please

**Comment by KrzysztofCwalina** [comment]
@dotnet-bot test this please

**Comment by adamsitnik** [comment]
You have edited SpanEqualsExtensions.cs which is auto-generated by SpanExtensionsTemplate.tt. IMHO you should edit tt file and regenerate the Extensions.

**Comment by KrzysztofCwalina** [comment]
@adamsitnik, ah good catch. i will fix that.

**Comment by KrzysztofCwalina** [comment]
@dotnet-bot test this please

**Comment by KrzysztofCwalina** [comment]
@dotnet-bot test this please

File: src/System.Slices/tests/Tests.cs [hunk 1]
```diff
@@ -47,12 +47,47 @@ public void TwoSpansCreatedOverSameIntArrayAreEqual()
...
+        }
+    }
+
     [Fact]
     public void TestEndsWith() {
         
         var str = "Hello, Slice!";
-        Span<char> slice = str.Slice();
-        Span<char> slice2 = "Slice!".Slice();
+        ReadOnlySpan<char> slice = str.Slice();
```
Comment (krwq) [inline src/System.Slices/tests/Tests.cs]: Should this test be copied first?

File: src/System.Slices/tests/Tests.cs [hunk 2]
```diff
@@ -62,7 +97,7 @@ public void TwoSpansCreatedOverSameIntArrayAreEqual()
     public void TwoSpansCreatedOverSameStringsAreEqual()
     {
         var str = "Hello, Slice!";
-        Span<char> slice = str.Slice();
+        ReadOnlySpan<char> slice = str.Slice();
```
Comment (krwq) [inline src/System.Slices/tests/Tests.cs]: same

File: src/System.Slices/tests/EqualityTests.cs [hunk 3]
```diff
@@ -136,19 +221,21 @@ public void SpansOfEqualValuesOfDifferentTypesAndDifferentSizesInSameOrderAreStr
         }
 
         [Fact]
-        public void AddressOfStringCanBePinnedToo()
+        public void ReadOnlySpanOfEqualValuesOfDifferentTypesAndDifferentSizesInSameOrderAreStructurallyEqual()
         {
-            string veryLongString = string.Join(string.Empty, Enumerable.Repeat("JustAVeryLongText", 1000));
-            string sameString = string.Join(string.Empty, Enumerable.Repeat("JustAVeryLongText", 1000));
+            var guids = Enumerable.Range(0, 10).Select(_ => Guid.NewGuid()).ToArray();
```
Comment (krwq) [inline src/System.Slices/tests/EqualityTests.cs]: is this accidental or intentional?

File: src/System.Slices/src/System/SpanExtensions.cs [hunk 4]
```diff
@@ -66,10 +66,10 @@ public static Span<T> Slice<T>(this T[] array, int start, int length)
         /// <exception cref="System.ArgumentException">
         /// Thrown if the 'str' parameter is null.
         /// </exception>
-        public static Span<char> Slice(this string str)
+        public static ReadOnlySpan<char> Slice(this string str)
```
Comment (krwq) [inline src/System.Slices/src/System/SpanExtensions.cs]: Should we have extensions for both?

Comment (krwq) [inline src/System.Slices/src/System/SpanExtensions.cs]: nevermind

File: src/System.Slices/tests/Tests.cs (hunk 1 above)
Comment (krwq) [inline src/System.Slices/tests/Tests.cs]: ok, got it now - lgtm

File: src/System.Slices/tests/Tests.cs (hunk 2 above)
Comment (krwq) [inline src/System.Slices/tests/Tests.cs]: got it - lgtm

**Comment by krwq** [comment]
Looks good to me assuming it's merging

**Pull request merged**